web/src/copy_messages.ts

```diff
@@ -119,54 +119,88 @@ function get_nearest_html_element(node: Node | null): Element | null {
+    }
+    if (is_within_math_block) {
+        // One might think that this will break in case of empty katex-display(s)
+        // being the start or end node which is/are created when we insert
+        // some extra newlines within a math block.
+        // However, is it not possible to select those empty katex-displays
+        // as per my observation on Chrome and Firefox.
+        if (is_range_start) {
+            range.setStart(selection_element.closest(".katex-display")!, 0);
+        } else {
+            // The offset 1 selects the only child of `.katex-display`
+            // which is `.katex`.
+            range.setEnd(selection_element.closest(".katex-display")!, 1);
+        }
+    } else {
+        if (is_range_start) {
+            range.setStart(selection_element.closest(".katex")!, 0);
+        } else {
+            // The offset 2 selects the two children of `.katex`
+            // namely `.katex-mathml` and `.katex-html`
+            range.setEnd(selection_element.closest(".katex")!, 2);
+        }
+    }
+}
 /*
     This is done to make the copying within math blocks smarter.
     We mutate the selection for selecting singular katex expressions
     within math blocks when applicable, which will be
     converted into the inline $$<expr>$$ syntax.
 
-    In case when a single expression or its subset is selected
-    within a math block, we adjust the selection so that it
-    selects the katex span which is parenting that expression.
-    This converts the selection into an inline expression as
-    per the turndown rules below.
-
-    We want to avoid this behavior if the selection
-    spreads across multiple katex displays i.e. the
-    focus and anchor are not part of the same katex span.
+    We expand the selection range only in the following cases:
+    1. Either the startContainer or endContainer or both are within an inline expression where the
+    range covers one or more math expressions.
+    2. Either the startContainer, endContainer, or both are within a math block where the
+    range covers one or more math expressions.
+
+    In principle, we only need to expand the start of the selection range for the
+    cases where multiple expressions are selected because the end of the range
+    always contains the annotation element in case it lies within the math block.
+
+    But, we still expand the end of the range to select the complete expression.
+    This is done in cases where the startContainer/endContainer denoting the end of the range
+    lies inside the math block or expression(depending on whether startContainer comes after the endContainer or
+    vice-versa in the DOM) as part of the effort to keep selection range and paste content
+    as close as possible, to eventually tackle #6316.
+
+    The primary objective of this function is to include the `<annotation>` tag within
+    the selection range so that it can be utilized later by `paste_handler_converter`
+    as `paste_html`.
```

Comment:
I rewrote this comment, since I don't think it's possible for `endContainer` to be before `startContainer`; that's likely stale text from when you were using focus/anchor. Also just tried to make it focus on what's important.